rust/automerge/src/op_tree.rs

```diff
@@ -71,78 +190,200 @@ impl OpTreeInternal {
+                succ.push(pos);
+            }
+
+            pos += 1;
+            next = iter.next();
         }
+
+        SeekOpFound::simple(pos, succ)
     }
 
-    pub(crate) fn keys_at(&self, clock: Clock) -> Option<query::KeysAt<'_>> {
-        if self.root_node.is_some() {
-            Some(query::KeysAt::new(self, clock))
-        } else {
-            None
+    pub(crate) fn seek_update_for_op_for_observer<'a>(
+        &'a self,
+        meta: &OpSetMetadata,
+        op: &'a Op,
+        mut pos: usize,
+        index: usize,
+    ) -> SeekOpFound<'a> {
+        let mut iter = self.iter();
+        let mut found = None;
+        let mut before = None;
+        let mut num_before = 0;
+        let mut overwritten = None;
+        let mut after = None;
+        let mut succ = vec![];
+        let mut next = iter.nth(pos);
+        while let Some(e) = next {
+            if e.elemid_or_key() != op.elemid_or_key() {
+                break;
+            }
+
+            if found.is_none() && meta.lamport_cmp(e.id, op.id) == Ordering::Greater {
+                found = Some(pos);
+            }
+
+            if op.overwrites(e) {
+                succ.push(pos);
+
+                if e.visible() {
+                    overwritten = Some(e);
+                }
+            } else if e.visible() {
+                if found.is_none() && overwritten.is_none() {
+                    before = Some(e);
+                    num_before += 1;
+                } else {
+                    after = Some(e);
+                }
+            }
+
+            pos += 1;
+            next = iter.next();
+        }
+
+        pos = found.unwrap_or(pos);
+
+        SeekOpFound {
+            before,
+            num_before,
+            after,
+            overwritten,
+            succ,
+            pos,
+            index,
         }
     }
 
-    pub(crate) fn map_range<'a, R: RangeBounds<String>>(
+    pub(crate) fn seek_opid<'a>(
         &'a self,
-        range: R,
-        meta: &'a OpSetMetadata,
-    ) -> Option<query::MapRange<'a, R>> {
-        if self.root_node.is_some() {
-            Some(query::MapRange::new(range, self, meta))
-        } else {
-            None
+        opid: OpId,
+        encoding: ListEncoding,
+        clock: Option<&Clock>,
+        meta: &OpSetMetadata,
+    ) -> Option<(&'a Op, usize, bool)> {
+        let query = self.search(query::OpIdSearch::opid(opid, encoding, clock), meta);
+        let pos = query.found()?;
+        let mut iter = self.iter();
+        let op = iter.nth(pos)?;
+        let index = query.index_for(op);
+        for e in iter {
+            if e.elemid_or_key() != op.elemid_or_key() {
+                break;
+            }
+
+            if e.visible() {
+                return Some((op, index, false));
+            }
         }
+        Some((op, index, op.visible()))
     }
 
-    pub(crate) fn map_range_at<'a, R: RangeBounds<String>>(
+    pub(crate) fn seek_op_with_observer<'a>(
```

Comment:
I would find this easier to read if it were `seek_to_op_with_observer` or `find_op_with_observer`